dango/lending-pool/src/execute.rs

```diff
@@ -0,0 +1,243 @@
+    ensure!(!ctx.funds.is_empty(), "No funds to withdraw");
+
+    let cfg = ctx.querier.query_config()?;
+    let mut msgs = vec![];
+    let mut withdrawn = Coins::new();
+    for coin in ctx.funds.into_iter() {
+        // Ensure only LP tokens are sent
+        let mut iter = coin.denom.inner().iter();
+
+        ensure!(
+            iter.next().map(|part| part.as_ref()) == Some(NAMESPACE),
+            "namespace:{NAMESPACE} not found"
+        );
+
+        ensure!(
+            iter.next().map(|part| part.as_ref()) == Some("lp"),
+            "namespace: lp not found"
+        );
+
+        // let denom_parts = iter.map(|part| part.clone()).collect::<Vec<_>>();
+
+        // Add msg to send the underlying tokens to the recipient
+        let underlying_denom = Denom::from_parts(iter.cloned().collect::<Vec<_>>())?;
+        let amount = coin.amount;
+        withdrawn.insert(Coin::new(underlying_denom, amount)?)?;
+
+        // Burn the LP tokens
+        msgs.push(Message::execute(
+            cfg.bank,
+            &bank::ExecuteMsg::Burn {
+                from: ctx.contract,
+                denom: coin.denom,
+                amount,
+            },
+            Coins::new(),
+        )?);
+    }
+
+    // Transfer the underlying tokens to the recipient
+    msgs.push(Message::transfer(recipient, withdrawn)?);
+
+    Ok(Response::new().add_messages(msgs))
+}
+
+pub fn borrow(ctx: MutableCtx, coins: Coins) -> anyhow::Result<Response> {
+    // Ensure sender is a margin account
+    ensure_sender_account_is_margin(&ctx)?;
+
+    // Ensure the coins are whitelisted
+    for coin in &coins {
+        ensure!(
+            WHITELISTED_DENOMS.has(ctx.storage, coin.denom.clone()),
+            "Invalid denom. Only whitelisted denoms can be borrowed."
+        );
+    }
+
+    // Update the sender's liabilities
+    LIABILITIES.may_update(ctx.storage, ctx.sender, |debts| {
+        let mut debts = debts.unwrap_or_default();
+        debts.insert_many(coins.clone())?;
+        Ok(debts)
+    })?;
+
+    // Transfer the coins to the caller
+    Ok(Response::new().add_message(Message::transfer(ctx.sender, coins)?))
+}
+
+// ----------------------------------- tests -----------------------------------
+
+#[cfg(test)]
+mod tests {
+
+    use {
+        super::*,
+        grug::{MockContext, MockQuerier},
+    };
+
+    /// Address of the Lending Pool for use in the following tests.
+    const LENDING_POOL: Addr = Addr::mock(255);
+
+    #[test]
+    fn cant_transfer_to_lending_pool() {
+        let querier = MockQuerier::new();
+        let mut ctx = MockContext::new()
+            .with_querier(querier)
+            .with_contract(LENDING_POOL)
+            .with_sender(Addr::mock(123))
+            .with_funds(Coins::new());
+
+        let res = receive(ctx.as_mutable());
+        assert!(res.is_err_and(|err| err
+            .to_string()
+            .contains("Can't send tokens to this contract")));
+    }
```

Comment:
Is this test necessary? You already have an integration that tests the same thing.

Reply:
Nah just did it first with unit test but prefer integration test.